src/main/java/com/typesafe/netty/http/pipelining/HttpPipeliningHandler.java

```diff
@@ -1,110 +1,221 @@
+ * message in response to the request, and the last, and only the last message it sends must have the last flag set.
+ *
+ * If messages are sent after the last message is sent, these messages may end up being buffered until the sequence
+ * number overflows and cycles back. If messages for a given sequence number are not sent sequentially, similar
+ * behaviour could also occur.
+ *
+ * There is no limit to the amount of messages that this handler will buffer for a particular sequence number.  It is
+ * the responsibility of the handler sending the outbound messages to handle back pressure via promises associated
+ * with each write event - if this is done, the buffering will be inherently bounded by back pressure.
  *
- * @author Christopher Hunt
+ * This handler does however put a bound on the maximum number of in flight requests that it will handle, configured by
+ * inFlightRequestsLowWatermark and inFlightRequestsHighWatermark. When the high watermark is exceeded, the handler will
+ * push back on the client. When the low watermark is reached, the handler will start reading again.  This back pressure
+ * mechanism only works if ChannelOptions.AUTO_READ is false.
+ *
+ * This back pressure is implemented by blocking channelReadComplete events, so assumes that the following handlers
+ * will not request reading unless they receive this event.  Note that the handler does nothing to actually block
+ * incoming requests when the high watermark is reached, it only pushes back on the TCP connection.  If there are more
+ * requests in the TCP buffers before this back pressure takes effect, including the current request that triggered the
+ * high water mark to be exceeded, these requests will still be sent to the following handlers.
+ *
+ * When the low watermark is reached after the high watermark has been exceeded, the handler will stop blocking
+ * channelReadComplete events, and will issue a read.
  */
-public class HttpPipeliningHandler extends SimpleChannelHandler {
+public class HttpPipeliningHandler extends ChannelDuplexHandler {
 
-    public static final int INITIAL_EVENTS_HELD = 3;
-    public static final int MAX_EVENTS_HELD = 10000;
+    /**
+     * The sequence of received HTTP requests.
+     */
+    volatile private int receiveSequence = 0;
 
-    private final int maxEventsHeld;
+    /**
+     * The currently sending sequence of HTTP requests.
+     */
+    volatile private int currentlySendingSequence = 1;
 
-    private int sequence;
-    private int nextRequiredSequence;
-    private int nextRequiredSubsequence;
+    /**
+     * Whether the high watermark has been exceeded.
+     */
+    volatile private boolean highWatermarkExceeded = false;
 
-    private final Queue<OrderedDownstreamChannelEvent> holdingQueue;
+    /**
+     * The number of requests in flight
+     */
+    private final AtomicInteger inFlight = new AtomicInteger();
```

Comment:
Updated - I actually removed inFlight since it can be determined from receive sequence and current sending sequence.